### PR TITLE
feat: add `reduce` and `reduceRight` methods to `array/bool`

### PR DESCRIPTION
**Pull request by Jaysukh-409**
Resolves: Subtask of #2304 

## Description

> What is the purpose of this pull request?

This pull request:

-   Adds the typed array `reduce` and `reduceRight` methods to prototype of `BooleanArray`

## Related Issues

> Does this pull request have any related issues?

This pull request:

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

File: lib/node_modules/@stdlib/array/bool/benchmark/benchmark.reduce.js [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+* Reducer function.
+* @private
+* @param {any} acc - accumulated value
+* @param {boolean} value - current array element
+* @param {integer} index - current array index
+* @returns {any} accumulated value
+*/
+function reducer( acc, value ) {
+	return Boolean( acc & value );
+}
```
Comment (Jaysukh-409) [inline lib/node_modules/@stdlib/array/bool/benchmark/benchmark.reduce.js]: Not sure whether it is good way for benchmarking.

File: lib/node_modules/@stdlib/array/bool/README.md [hunk 2]
```diff
@@ -779,6 +779,106 @@ var count = context.count;
...
+#### BooleanArray.prototype.reduce( reducerFn\[, initialValue] )
+
+Applies a provided callback function to each element of the array, in order, passing in the return value from the calculation on the preceding element and returning the accumulated result upon completion.
+
+```javascript
+var Boolean = require( '@stdlib/boolean/ctor' );
+
+function reducer( acc, v ) {
+    return Boolean( acc & v );
+}
```
Comment (Jaysukh-409) [inline lib/node_modules/@stdlib/array/bool/README.md]: Can we remove Boolean constructor from here, by return `acc && v` ?

File: lib/node_modules/@stdlib/array/bool/README.md [hunk 3]
```diff
@@ -779,6 +779,106 @@ var count = context.count;
...
+    return acc;
+}
+
+var arr = new BooleanArray( 3 );
+
+arr.set( true, 0 );
+arr.set( false, 1 );
+arr.set( true, 2 );
+
+var sum = arr.reduce( reducer );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/array/bool/README.md]:
```suggestion
var sum = arr.reduce( reducer, 0 );
```

File: lib/node_modules/@stdlib/array/bool/README.md (hunk 2 above)
Comment (kgryte) [reply lib/node_modules/@stdlib/array/bool/README.md]: You don't need the boolean constructor and can remove.

File: lib/node_modules/@stdlib/array/bool/README.md [hunk 4]
```diff
@@ -779,6 +779,106 @@ var count = context.count;
...
+
+#### BooleanArray.prototype.reduce( reducerFn\[, initialValue] )
+
+Applies a provided callback function to each element of the array, in order, passing in the return value from the calculation on the preceding element and returning the accumulated result upon completion.
+
+```javascript
+var Boolean = require( '@stdlib/boolean/ctor' );
+
+function reducer( acc, v ) {
+    return Boolean( acc & v );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/array/bool/README.md]:
```suggestion
    return Boolean( acc && v );
```

I don't think you want to use `&` (bitwise AND) and instead want to use logical AND.

Comment (kgryte) [reply lib/node_modules/@stdlib/array/bool/README.md]: Same thing. You don't need the `Boolean` constructor here and can remove.

File: lib/node_modules/@stdlib/array/bool/README.md [hunk 5]
```diff
@@ -779,6 +779,106 @@ var count = context.count;
...
+```
+
+<a name="method-reduce-right"></a>
+
+#### Complex64Array.prototype.reduceRight( reducerFn\[, initialValue] )
+
+Applies a provided callback function to each element of the array, in reverse order, passing in the return value from the calculation on the following element and returning the accumulated result upon completion.
+
+```javascript
+var Boolean = require( '@stdlib/boolean/ctor' );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/array/bool/README.md]: Same comments as above.

File: lib/node_modules/@stdlib/array/bool/README.md [hunk 6]
```diff
@@ -779,6 +779,106 @@ var count = context.count;
...
+    return acc;
+}
+
+var arr = new BooleanArray( 3 );
+
+arr.set( true, 0 );
+arr.set( false, 1 );
+arr.set( true, 2 );
+
+var sum = arr.reduceRight( reducer );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/array/bool/README.md]:
```suggestion
var sum = arr.reduceRight( reducer, 0 );
```

Missing the initial value, which is what this example is supposed to illustrate.

File: lib/node_modules/@stdlib/array/bool/benchmark/benchmark.reduce.js (hunk 1 above)
Comment (kgryte) [reply lib/node_modules/@stdlib/array/bool/benchmark/benchmark.reduce.js]: I'd use the "sum" example from the README. Also, make sure you understand the difference between `&` and `&&`. They are not the same thing.

File: lib/node_modules/@stdlib/array/bool/benchmark/benchmark.reduce.length.js [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+var Boolean = require( '@stdlib/boolean/ctor' );
+var pkg = require( './../package.json' ).name;
+var BooleanArray = require( './../lib' );
+
+
+// FUNCTIONS //
+
+/**
+* Reducer function.
+* @private
```
Comment (kgryte) [inline lib/node_modules/@stdlib/array/bool/benchmark/benchmark.reduce.length.js:36]:
```suggestion
*
* @private
```

File: lib/node_modules/@stdlib/array/bool/benchmark/benchmark.reduce.length.js [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+* @param {any} acc - accumulated value
+* @param {boolean} value - current array element
+* @param {integer} index - current array index
+* @returns {any} accumulated value
+*/
+function reducer( acc, value ) {
+	return Boolean( acc & value );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/array/bool/benchmark/benchmark.reduce.length.js]: Same thing. Use sum reducer.

File: lib/node_modules/@stdlib/array/bool/benchmark/benchmark.reduce.length.js [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+var pkg = require( './../package.json' ).name;
+var BooleanArray = require( './../lib' );
+
+
+// FUNCTIONS //
+
+/**
+* Reducer function.
+* @private
+* @param {any} acc - accumulated value
```
Comment (kgryte) [inline lib/node_modules/@stdlib/array/bool/benchmark/benchmark.reduce.length.js]: We don't use `any` in JSDoc.

File: lib/node_modules/@stdlib/array/bool/benchmark/benchmark.reduce.length.js [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+
+		b.tic();
+		for ( i = 0; i < b.iterations; i++ ) {
+			out = arr.reduce( reducer );
+			if ( typeof out !== 'boolean' ) {
+				b.fail( 'should return a boolean' );
+			}
+		}
+		b.toc();
+		if ( !isBoolean( out ) ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/array/bool/benchmark/benchmark.reduce.length.js]: When refactoring to compute a sum, you'll need to update these checks.

File: lib/node_modules/@stdlib/array/bool/benchmark/benchmark.reduce_right.js [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+* Reducer function.
+* @private
+* @param {any} acc - accumulated value
+* @param {boolean} value - current array element
+* @param {integer} index - current array index
+* @returns {any} accumulated value
+*/
+function reducer( acc, value ) {
+	return Boolean( acc & value );
+}
```
Comment (kgryte) [inline lib/node_modules/@stdlib/array/bool/benchmark/benchmark.reduce_right.js]: Same comments as other benchmark file.

File: lib/node_modules/@stdlib/array/bool/benchmark/benchmark.reduce_right.length.js [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+* Reducer function.
+* @private
+* @param {any} acc - accumulated value
+* @param {boolean} value - current array element
+* @param {integer} index - current array index
+* @returns {any} accumulated value
+*/
+function reducer( acc, value ) {
+	return Boolean( acc & value );
+}
```
Comment (kgryte) [inline lib/node_modules/@stdlib/array/bool/benchmark/benchmark.reduce_right.length.js]: Same comments.

File: lib/node_modules/@stdlib/array/bool/docs/repl.txt [hunk 13]
```diff
@@ -612,6 +612,90 @@
...
+    -------
+    out: any
+        Accumulated result.
+
+    Examples
+    --------
+    > function reducer( acc, v ) { return (acc &&  v); };
+    > var arr = new {{alias}}( [ true, false, true ] )
+    <BooleanArray>
+    > var sum = arr.reduce( reducer )
```
Comment (kgryte) [inline lib/node_modules/@stdlib/array/bool/docs/repl.txt]: This isn't a "sum", per se. I'd rename `sum` to `out`.

File: lib/node_modules/@stdlib/array/bool/docs/repl.txt [hunk 14]
```diff
@@ -612,6 +612,90 @@
...
+    -------
+    out: any
+        Accumulated result.
+    
+    Examples
+    --------
+    > function reducer( acc, v ) { return (acc &&  v); };
+    > var arr = new {{alias}}( [ true, false, true ] )
+    <BooleanArray>
+    > var sum = arr.reduceRight( reducer )
```
Comment (kgryte) [inline lib/node_modules/@stdlib/array/bool/docs/repl.txt]: Same comment.

File: lib/node_modules/@stdlib/array/bool/test/test.reduce.js [hunk 15]
```diff
@@ -0,0 +1,218 @@
...
+	t.end();
+});
+
+tape( 'attached to the prototype of the main export is a `reduce` method', function test( t ) {
+	t.strictEqual( hasOwnProp( BooleanArray.prototype, 'reduce' ), true, 'has property' );
+	t.strictEqual( isFunction( BooleanArray.prototype.reduce ), true, 'has method' );
+	t.end();
+});
+
+tape( 'the method throws an error if invoked with a `this` context which is not a complex number array instance', function test( t ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/array/bool/test/test.reduce.js]: copy-paste error

File: lib/node_modules/@stdlib/array/bool/test/test.reduce.js [hunk 16]
```diff
@@ -0,0 +1,218 @@
...
+		function noop() {}
+	];
+	for ( i = 0; i < values.length; i++ ) {
+		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
+	}
+	t.end();
+
+	function badValue( value ) {
+		return function badValue() {
+			return arr.map.reduce( value, reducer );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/array/bool/test/test.reduce.js]:
```suggestion
			return arr.reduce.call( value, reducer );
```

File: lib/node_modules/@stdlib/array/bool/test/test.reduce.js [hunk 17]
```diff
@@ -0,0 +1,218 @@
...
+	t.end();
+
+	function badValue( value ) {
+		return function badValue() {
+			return arr.reduce( value );
+		};
+	}
+});
+
+tape( 'the method throws an error if not provided an initial value when operating on an empty complex number array', function test( t ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/array/bool/test/test.reduce.js]: Copy paste error.

File: lib/node_modules/@stdlib/array/bool/test/test.reduce.js [hunk 18]
```diff
@@ -0,0 +1,218 @@
...
+
+	t.strictEqual( actual, expected, 'returns expected value' );
+
+	t.end();
+
+	function reducer( acc, value, index ) {
+		ind = index-1;
+		t.strictEqual( acc, accArray[ ind ], 'returns expected value' );
+		t.strictEqual( value, valueArray[ ind ], 'returns expected value' );
+		return Boolean( acc & value );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/array/bool/test/test.reduce.js]: `&` v `&&`

Comment (kgryte) [reply lib/node_modules/@stdlib/array/bool/test/test.reduce.js]: You also don't need the `Boolean` constructor here for what you are wanting to do.

File: lib/node_modules/@stdlib/array/bool/test/test.reduce.js [hunk 19]
```diff
@@ -0,0 +1,218 @@
...
+
+	arr = new BooleanArray( [ true, false, true, true ] );
+	expected = false;
+	actual = arr.reduce( reducer );
+
+	t.strictEqual( actual, expected, 'returns expected value' );
+	t.end();
+
+	function reducer( acc, value ) {
+		return Boolean( acc & value );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/array/bool/test/test.reduce.js]: `&` vs `&&`.

Comment (kgryte) [reply lib/node_modules/@stdlib/array/bool/test/test.reduce.js]: No need for `Boolean` constructor here and elsewhere.

File: lib/node_modules/@stdlib/array/bool/test/test.reduce.js [hunk 20]
```diff
@@ -0,0 +1,218 @@
...
+	t.strictEqual( actual, expected, 'returns expected value' );
+	t.end();
+
+	function reducer( acc, value ) {
+		if ( value ) {
+			return acc + 1;
+		}
+		return acc;
+	}
+});
```
Comment (kgryte) [inline lib/node_modules/@stdlib/array/bool/test/test.reduce.js]: This test is not applicable and was copy pasted from complex array tests.

File: lib/node_modules/@stdlib/array/bool/test/test.reduce_right.js [hunk 21]
```diff
@@ -0,0 +1,221 @@
...
+	t.end();
+});
+
+tape( 'attached to the prototype of the main export is a `reduceRight` method', function test( t ) {
+	t.strictEqual( hasOwnProp( BooleanArray.prototype, 'reduceRight' ), true, 'has property' );
+	t.strictEqual( isFunction( BooleanArray.prototype.reduceRight ), true, 'has method' );
+	t.end();
+});
+
+tape( 'the method throws an error if invoked with a `this` context which is not a complex number array instance', function test( t ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/array/bool/test/test.reduce_right.js]: All the same comments as the other test file.

**Review by kgryte — CHANGES_REQUESTED** [review]
This PR needs a good bit of clean-up. All comments should be straightforward to address.

File: lib/node_modules/@stdlib/array/bool/docs/repl.txt (hunk 13 above)
Comment (kgryte) [inline lib/node_modules/@stdlib/array/bool/docs/repl.txt]:
```suggestion
    > var out = arr.reduce( reducer )
```

File: lib/node_modules/@stdlib/array/bool/docs/repl.txt [hunk 22]
```diff
@@ -612,6 +612,90 @@
...
+        Initial accumulation value.
+
+    Returns
+    -------
+    out: any
+        Accumulated result.
+
+    Examples
+    --------
+    > function reducer( acc, v ) { return (acc &&  v); };
```
Comment (kgryte) [inline lib/node_modules/@stdlib/array/bool/docs/repl.txt]:
```suggestion
    > function reducer( acc, v ) { return ( acc && v ); };
```

File: lib/node_modules/@stdlib/array/bool/docs/repl.txt [hunk 23]
```diff
@@ -612,6 +612,90 @@
...
+        Initial accumulation value.
+    
+    Returns
+    -------
+    out: any
+        Accumulated result.
+    
+    Examples
+    --------
+    > function reducer( acc, v ) { return (acc &&  v); };
```
Comment (kgryte) [inline lib/node_modules/@stdlib/array/bool/docs/repl.txt]:
```suggestion
    > function reducer( acc, v ) { return ( acc && v ); };
```

File: lib/node_modules/@stdlib/array/bool/docs/types/index.d.ts [hunk 24]
```diff
@@ -516,6 +570,58 @@ declare class BooleanArray implements BooleanArrayInterface {
...
+	*
+	* @param reducer - callback function
+	* @param initialValue - initial value
+	* @returns accumulated result
+	*
+	* @example
+	* function reducer( acc, v ) {
+	* 	  if ( v ) {
+	*	       return acc + 1;
+	*	  }
```
Comment (kgryte) [inline lib/node_modules/@stdlib/array/bool/docs/types/index.d.ts]:
```suggestion
	*     if ( v ) {
	*          return acc + 1;
	*     }
```

File: lib/node_modules/@stdlib/array/bool/docs/types/index.d.ts [hunk 25]
```diff
@@ -516,6 +570,58 @@ declare class BooleanArray implements BooleanArrayInterface {
...
```
Comment (kgryte) [inline lib/node_modules/@stdlib/array/bool/docs/types/index.d.ts]:
```suggestion
	*     if ( v ) {
	*          return acc + 1;
	*     }
```

File: lib/node_modules/@stdlib/array/bool/lib/main.js [hunk 26]
```diff
@@ -978,6 +978,124 @@ setReadOnly( BooleanArray.prototype, 'map', function map( fcn, thisArg ) {
...
+* @param {*} [initialValue] - initial value
+* @throws {TypeError} `this` must be a boolean array
+* @throws {Error} if not provided an initial value, the array must have at least one element
+* @returns {*} accumulated result
+*
+* @example
+* function reducer( acc, v ) {
+* 	  if ( v ) {
+*	       return acc + 1;
+*	  }
```
Comment (kgryte) [inline lib/node_modules/@stdlib/array/bool/lib/main.js]:
```suggestion
*     if ( v ) {
*          return acc + 1;
*     }
```

File: lib/node_modules/@stdlib/array/bool/lib/main.js [hunk 27]
```diff
@@ -978,6 +978,124 @@ setReadOnly( BooleanArray.prototype, 'map', function map( fcn, thisArg ) {
...
+*     return acc;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/array/bool/lib/main.js]:
```suggestion
*     if ( v ) {
*          return acc + 1;
*     }
*     return acc;
```

**Review by kgryte — APPROVED** [review]
LGTM. Thanks, @Jaysukh-409!

**Pull request merged**